### PR TITLE
Implement exercise binary-search-tree

### LINKED ISSUE #740: binary-search-tree: implement exercise
There are exercises that are not yet implemented in Python track (the full list of them - https://github.com/exercism/python/issues/417#issuecomment-334841778).
You can help us change this situation by implementing this one. Ref #417

The guide ["Implement an exercise from specification"](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md) will help you succeed.

If you still have questions, I encourage you to ask them before starting working on the issue.

I also suggest claiming the issue by creating an empty PR as stated [here](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md#avoiding-duplicate-work)

Please note that there are some special ['closing issue' keywords](https://help.github.com/articles/closing-issues-using-keywords/) that you can put in description of your PR and Github will take care of closing the corresponding issue automatically after merge.
This simple thing helps us a lot, and makes our life a bit easier

### PR DESCRIPTION
**Pull request by rodrigocam**
Resolved #740

### CODE REVIEW

**Comment by m-a-ge** [comment]
You've created two PRs #773 and #768 
Could you please close one to avoid confusion

File: exercises/binary-search-tree/example.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                        return inserted
+                elif(value > cur_node.value):
+                    if(cur_node.right_node):
+                        cur_node = cur_node.right_node
+                    else:
+                        cur_node.right_node = TreeNode(value)
+                        inserted = True
+                        return inserted
+
+    def search(self, searched_number):
```
Comment (kabiir) [inline exercises/binary-search-tree/example.py]: I recommend changing searched_number to either `number` or `search_number`, searched is in past tense which means you've already searched for it.

File: exercises/binary-search-tree/example.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        cur_node = self.root
```
Comment (kabiir) [inline exercises/binary-search-tree/example.py]: Also add a check for `if(self.root is None): return false` which does an early exit if the tree hasn't been constructed.

File: exercises/binary-search-tree/example.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    def print_preorder(self, node):
+        if(node is not None):
+            print(node.value)
+            self.print_preorder(node.left_node)
+            self.print_preorder(node.right_node)
+
+    def print_postorder(self, node):
+        if(node is not None):
+            self.print_preorder(node.left_node)
```
Comment (kabiir) [inline exercises/binary-search-tree/example.py]: Post order should be
```python
self.print_postorder(node.right_node) # Right comes first
self.print_postorder(node.left_node)
print(node.value)
```
Also you're calling `print_preorder()` .

File: exercises/binary-search-tree/example.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                        cur_node.right_node = TreeNode(value)
+                        inserted = True
+                        return inserted
+
+    def search(self, searched_number):
+        cur_node = self.root
+        found = False
+        while not found:
+            if(cur_node is None):
+                raise ValueError("Element not found")
```
Comment (kabiir) [inline exercises/binary-search-tree/example.py]: You won't need to raise an exception if you do an early `return false`

File: exercises/binary-search-tree/example.py (hunk 1 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/example.py]: Ok, changed!

File: exercises/binary-search-tree/example.py (hunk 2 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/example.py]: Ok, checked!

File: config.json [hunk 5]
```diff
@@ -888,9 +888,22 @@
       "uuid": "6b7f7b1f-c388-4f4c-b924-84535cc5e1a0",
       "slug": "hexadecimal",
       "deprecated": true
+    },
```
Comment (cmccandless) [inline config.json]: It would probably be best to move the `config.json` entry for this exercise before all deprecated exercises.

File: exercises/binary-search-tree/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+Insert and search for numbers in a binary tree.
```
Comment (N-Parsons) [inline exercises/binary-search-tree/README.md]: This needs the title of the exercise as a heading: `# Binary Search Tree`

File: exercises/binary-search-tree/binary_search_tree.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class TreeNode(object):
+    def __init__(self, value):
+        pass
+
+
+class BinarySearchTree(object):
+    def __init__(self):
+        pass
```
Comment (N-Parsons) [inline exercises/binary-search-tree/binary_search_tree.py]: Can you also include skeletons for the `add` and `search` methods here?

File: exercises/binary-search-tree/binary_search_tree.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class TreeNode(object):
+    def __init__(self, value):
+        pass
```
Comment (N-Parsons) [inline exercises/binary-search-tree/binary_search_tree.py]: Can you add `self.value = value` here? I feel like if we assume that something exists in the tests, then we should make it clear in the solution template.

File: config.json [hunk 9]
```diff
@@ -1169,6 +1183,5 @@
     }
   ],
   "foregone": [
-
```
Comment (N-Parsons) [inline config.json]: Can you add this blank line back in?

File: exercises/binary-search-tree/README.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      3
+
+And if we then added 1, 5, and 7, it would look like this
+
+          4
+        /   \
+       /     \
+      2       6
+     / \     / \
+    1   3   5   7
```
Comment (N-Parsons) [inline exercises/binary-search-tree/README.md]: There should also be some other stuff that goes here. Check another exercise and copy it over, or regenerate this README with the `configlet`.

File: exercises/binary-search-tree/binary_search_tree_test.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import unittest
+
+from binary_search_tree import BinarySearchTree
+
+
+class BinarySearchTreeTests(unittest.TestCase):
+    bst = BinarySearchTree()
+
+    def test_add_integer_number1(self):
+        self.assertTrue(self.bst.add(7))
```
Comment (cmccandless) [inline exercises/binary-search-tree/binary_search_tree_test.py]: We use `assertIs(actual, True)` and `assertIs(actual, False)` rather than `assertTrue(actual)` or `assertFalse(actual)` ([#419](https://github.com/exercism/python/pull/419)).

File: exercises/binary-search-tree/example.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                        cur_node.left_node = TreeNode(value)
+                        inserted = True
+                        return inserted
+                elif(value > cur_node.value):
+                    if(cur_node.right_node):
+                        cur_node = cur_node.right_node
+                    else:
+                        cur_node.right_node = TreeNode(value)
+                        inserted = True
+                        return inserted
```
Comment (N-Parsons) [inline exercises/binary-search-tree/example.py]: Looking at this again, I just noticed that this method is returning `True`/`False` to indicate success/failure. Under what condition is failure expected to occur? From what I can see, this code will never return `False`, and would actually just enter an infinite loop if `value == cur_node.value`.

I think it would be worth adding a `BinarySearchTree.list` method to this exercise that returns a sorted list of values in the tree. A `collections.deque` would be most efficient for this, but I don't think that the tests should enforce its use, so the tests should probably convert the actual output to a list (eg. `assertEqual(list(self.bst.list()), [1, 2, 3])`).

Doing this would allow us to check that values are placed in the right order - each test for adding a number should contain a second assert for the current list of stored values. We should also test duplicate values, because the README specifies a way for these to be handled.

**From the `README`:**
> All data in the left subtree is less than or equal to the current node's data, and all data in the right subtree is greater than the current node's data.

Comment (cmccandless) [reply exercises/binary-search-tree/example.py]: @N-Parsons Returning `True` from an add or insert method is a common practice for container classes. I agree that it doesn't make the most sense, but that is likely why @rodrigocam did this.

**Comment by cmccandless** [comment]
@rodrigocam Just FYI, canonical-data for this exercise is being created at [/exercism/problem-specifications#940](https://github.com/exercism/problem-specifications/pull/940).

File: config.json (hunk 9 above)
Comment (rodrigocam) [reply config.json]: Done

File: exercises/binary-search-tree/README.md (hunk 6 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/README.md]: Done

File: exercises/binary-search-tree/README.md (hunk 10 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/README.md]: Done

File: exercises/binary-search-tree/binary_search_tree.py (hunk 8 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree.py]: Done

File: exercises/binary-search-tree/binary_search_tree.py (hunk 7 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree.py]: Done

File: exercises/binary-search-tree/binary_search_tree_test.py (hunk 11 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree_test.py]: Done

File: exercises/binary-search-tree/example.py (hunk 12 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/example.py]: @cmccandless is right, I thought this could be good to give more safety. I made the list method that @N-Parsons asked and change the tests to use this new method. I fix the infinite loop condition that you point but I keep the bool return, if it was a problem I can remove (:

Comment (N-Parsons) [reply exercises/binary-search-tree/example.py]: Given that there's no way that `False` can be returned, and there is not test for this method returning `False`, I think it's unnecessary and likely to cause some confusion/frustration. I think it's fine for this example code to return `True`/`False`, but I don't think that it should be tested for.

File: exercises/binary-search-tree/binary_search_tree_test.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        bst.add(7.5)
+        self.assertEqual(bst.search(1).value, 1)
+        self.assertEqual(bst.search(7.5).value, 7.5)
+        
+    def test_search_invalid_numbers(self):
+        bst = BinarySearchTree()
+        bst.add(1)
+        bst.add(7.5)
+        self.assertIs(bst.search(6), False)
+        self.assertIs(bst.search(8.8), False)
```
Comment (N-Parsons) [inline exercises/binary-search-tree/binary_search_tree_test.py]: I don't like that `search` will sometimes return a `bool` and sometimes return a `TreeNode`. I think that this should either be `TreeNode`/`None`, or `True`/`False` - the former is perhaps more generally useful, since nodes are truthy and `None` is falsey, and returning the actual node could be useful in some instances.

Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree_test.py]: Done! I change to return None when the node is not found.

File: exercises/binary-search-tree/example.py (hunk 12 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/example.py]: Done! I take out the return of the method.

File: config.json (hunk 9 above)
Comment (N-Parsons) [reply config.json]: I looks like you've got some unnecessary white-space here now (see below) - can you please remove it?

File: exercises/binary-search-tree/binary_search_tree.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class BinarySearchTree(object):
+    def __init__(self):
+        pass
+
+    def add(self, value):
+        pass
+
+    def search(self, search_number):
```
Comment (N-Parsons) [inline exercises/binary-search-tree/binary_search_tree.py]: I think this should use `value` rather than `search_number` as an argument, since we put values in and then search for them, and `search_` is probably redundant.

File: exercises/binary-search-tree/.cache/v/cache/lastfailed [hunk 15]
```diff
@@ -0,0 +1,10 @@
+{
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_add_float_number": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_add_integer_number1": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_add_integer_number2": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_add_integer_number3": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_add_integer_number4": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_search_valid_number1": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_search_valid_number2": true,
+  "binary_search_tree_test.py::BinarySearchTreeTests::test_search_valid_number3": true
+}
```
Comment (N-Parsons) [inline exercises/binary-search-tree/.cache/v/cache/lastfailed]: You need to remove the `.cache` directory.

**Comment by N-Parsons** [comment]
Hi @rodrigocam, sorry for taking a while to get back to this - I've had a really busy couple of weeks!

There are a couple of minor issues remaining that need to be dealt with, and then this can be merged.

File: exercises/binary-search-tree/binary_search_tree.py (hunk 14 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree.py]: Done!

File: exercises/binary-search-tree/.cache/v/cache/lastfailed (hunk 15 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/.cache/v/cache/lastfailed]: Done!

Comment (rodrigocam) [reply exercises/binary-search-tree/.cache/v/cache/lastfailed]: Done!

File: exercises/binary-search-tree/binary_search_tree.py (hunk 14 above)
Comment (N-Parsons) [reply exercises/binary-search-tree/binary_search_tree.py]: @rodrigocam, it looks like you've updated this in `example.py` but not here.

File: config.json [hunk 16]
```diff
@@ -1185,6 +1199,6 @@
...
+
```
Comment (N-Parsons) [inline config.json]: @rodrigocam, the extra spaces here also need to be removed. There should be a blank line, but there should be nothing in it.

**Comment by N-Parsons** [comment]
@cmccandless, can you review this again? There are a couple of minor changes pending and then I think this will be ready to merge.

File: exercises/binary-search-tree/binary_search_tree.py (hunk 14 above)
Comment (rodrigocam) [reply exercises/binary-search-tree/binary_search_tree.py]: Ok, done!

File: exercises/binary-search-tree/example.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        self.trav_inorder(self.root, elements)
+        return elements
+
+    def trav_inorder(self, node, elements):
+        if(node is not None):
+            self.trav_inorder(node.left_node, elements)
+            elements.append(node.value)
+            self.trav_inorder(node.right_node, elements)
+
+    def print_inorder(self, node):
```
Comment (cmccandless) [inline exercises/binary-search-tree/example.py]: We should probably remove all `print*` methods, as they are not used by the tests.

Comment (rodrigocam) [reply exercises/binary-search-tree/example.py]: Done!

**Comment by N-Parsons** [comment]
Looks good. Thanks @rodrigocam!
